File: apps/android/app/src/main/java/com/sanvya/app/ui/payments/PaymentHandleViewModel.kt
package com.sanvya.app.ui.payments

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sanvya.app.data.repository.HandleDisclosure
import com.sanvya.app.data.repository.SettingsRepository
import com.sanvya.app.data.repository.UpiRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

/**
 * "Your UPI ID" -- the Settings card's state.
 *
 * Ports `apps/web/src/payments/PaymentHandlePanel.tsx`. Every call belongs to
 * [UpiRepository]; what is here is web's four pieces of component state (the
 * saved hint, the busy flag, the error and the expanded log) plus the one
 * effect that loads the hint.
 *
 * Self-contained rather than fed from `SettingsViewModel`, matching
 * `SecurityViewModel`: web's panel calls `useSession()` itself, and reading the
 * session here keeps the card droppable into any screen the way web's is.
 *
 * The error string is the SERVER's, verbatim, and deliberately not an i18n key.
 * Every message this can surface comes from the `payment-handle` Edge Function
 * ("Create an account before saving a UPI ID.", "Couldn't save that: ..."), and
 * web renders `e.message` for exactly the same reason. Translating them would
 * mean maintaining a native copy of the function's error vocabulary that goes
 * stale silently. The strings the PANEL owns are all catalogued.
 *
 * Mirrors iOS's PaymentHandleViewModel.
 */
class PaymentHandleViewModel : ViewModel(), KoinComponent {

    private val upiRepository: UpiRepository by inject()
    private val settingsRepository: SettingsRepository by inject()

    /**
     * True until BOTH the session and the saved hint are known.
     *
     * Web's panel starts `loading = true` and its `useSession()` hydrates from a
     * localStorage cache, so it effectively never renders the guest message at
     * someone who is signed in. Native has no such cache, so the session read is
     * folded into the same loading flag -- otherwise every open of Settings
     * would flash "Create an account to add a UPI ID" at an account holder,
     * which is the same lie as flashing the empty form at someone who already
     * saved a handle (web's own comment on that effect).
     */
    private val _loading = MutableStateFlow(true)
    val loading: StateFlow<Boolean> = _loading.asStateFlow()

    /** Web's `useCanSavePaymentHandle()` -- signed in, and not a guest. */
    private val _canSave = MutableStateFlow(false)
    val canSave: StateFlow<Boolean> = _canSave.asStateFlow()

    /** The masked hint for the saved handle, or null when there isn't one. */
    private val _hint = MutableStateFlow<String?>(null)
    val hint: StateFlow<String?> = _hint.asStateFlow()

    private val _busy = MutableStateFlow(false)
    val busy: StateFlow<Boolean> = _busy.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    /**
     * Who has fetched your UPI ID. Synced, so it reads from local SQLite and
     * works offline -- see [UpiRepository.watchDisclosures].
     */
    val disclosures: StateFlow<List<HandleDisclosure>> = upiRepository.watchDisclosures()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    /**
     * The display name sent with a save, so the payer's UPI app has something
     * to show. Web passes `session?.username`.
     */
    private var username: String = ""

    private var started = false

    /**
     * Web's `useEffect(..., [canSave])`. Guarded so returning to Settings does
     * not refetch: the hint is a masked string that only changes when this
     * panel itself changes it.
     */
    fun start() {
        if (started) return
        started = true
        viewModelScope.launch {
            val session = try {
                settingsRepository.currentSession()
            } catch (_: Exception) {
                null
            }
            username = session?.username.orEmpty()
            // Bound through a safe call rather than smart-cast: `isGuest` is a
            // public property on a `:data` class, and Kotlin refuses to
            // smart-cast one declared in another module (PARITY_AUDIT trap 3,
            // the same one that bit Loans and Insights). Web's
            // `useCanSavePaymentHandle()` is `!!session && !session.isGuest`,
            // and no session means no save either way.
            val isGuest = session?.isGuest ?: true
            val allowed = session != null && !isGuest
            _canSave.value = allowed
            if (!allowed) {
                _loading.value = false
                return@launch
            }
            _hint.value = try {
                upiRepository.getMyPaymentHandle()
            } catch (_: Exception) {
                // The repository already degrades to the cached hint on a failed
                // read; anything that still throws is not worth an error banner
                // over a field the user has not touched yet.
                upiRepository.getCachedHint()
            }
            _loading.value = false
        }
    }

    /**
     * Web's `save()`. [normalizedVpa] is already trimmed and lower-cased.
     *
     * [onSaved] clears the input, which web does with `setValue("")` inside the
     * same `try`. The field lives in the composable, so the signal travels back
     * rather than the state moving up: the panel is otherwise stateless and a
     * `TextFieldValue` in a view model would fight `rememberSaveable` on
     * rotation.
     */
    fun save(normalizedVpa: String, onSaved: () -> Unit) {
        if (_busy.value) return
        _busy.value = true
        _error.value = null
        viewModelScope.launch {
            try {
                _hint.value = upiRepository.savePaymentHandle(normalizedVpa, username)
                onSaved()
            } catch (e: Exception) {
                _error.value = e.message
            } finally {
                _busy.value = false
            }
        }
    }

    /** Web's `forget()`. */
    fun forget() {
        if (_busy.value) return
        _busy.value = true
        _error.value = null
        viewModelScope.launch {
            try {
                upiRepository.forgetPaymentHandle()
                _hint.value = null
            } catch (e: Exception) {
                _error.value = e.message
            } finally {
                _busy.value = false
            }
        }
    }
}

File: apps/ios/App/ViewModels/PaymentHandleViewModel.swift
import Data
import Domain
import Factory
import Foundation
import Observation
import Supabase

/**
 "Your UPI ID" — the Settings section's state.

 Ports `apps/web/src/payments/PaymentHandlePanel.tsx`. Every call belongs to
 `UpiRepository`; what is here is web's four pieces of component state (the
 saved hint, the busy flag, the error and the expanded log) plus the one effect
 that loads the hint.

 Self-contained rather than fed from `SettingsViewModel`, matching
 `SecurityViewModel`: web's panel calls `useSession()` itself, and reading the
 session here keeps the section droppable into any screen the way web's is. The
 session read goes through `client.auth` for the same reason `SettingsViewModel`
 does — `AuthRepository` exposes `currentUserId` and `isGuest()` but not the
 username, and the username is what a payer's UPI app shows.

 The error string is the SERVER's, verbatim, and deliberately not an i18n key.
 Every message this can surface comes from the `payment-handle` Edge Function
 ("Create an account before saving a UPI ID.", "Couldn't save that: …"), and web
 renders `e.message` for exactly the same reason. Translating them would mean
 maintaining a native copy of the function's error vocabulary that goes stale
 silently. The strings the SECTION owns are all catalogued.

 Mirrors Android's PaymentHandleViewModel.kt.
 */
@Observable
@MainActor
final class PaymentHandleViewModel {

    @ObservationIgnored
    @Injected(\.upiRepository) private var upiRepository
    @ObservationIgnored
    @Injected(\.supabaseClient) private var client

    /**
     True until BOTH the session and the saved hint are known.

     Web's panel starts `loading = true` and its `useSession()` hydrates from a
     localStorage cache, so it effectively never renders the guest message at
     someone who is signed in. Native has no such cache, so the session read is
     folded into the same loading flag — otherwise every open of Settings would
     flash "Create an account to add a UPI ID" at an account holder, which is
     the same lie as flashing the empty form at someone who already saved a
     handle (web's own comment on that effect).
     */
    private(set) var loading = true

    /// Web's `useCanSavePaymentHandle()` — signed in, and not a guest.
    private(set) var canSave = false

    /// The masked hint for the saved handle, or nil when there isn't one.
    private(set) var hint: String?

    private(set) var busy = false
    private(set) var error: String?

    /// Who has fetched your UPI ID. Synced, so it reads from local SQLite and
    /// works offline — see `UpiRepository.watchDisclosures()`.
    private(set) var disclosures: [HandleDisclosure] = []

    /// The display name sent with a save, so the payer's UPI app has something
    /// to show. Web passes `session?.username`.
    @ObservationIgnored
    private var username = ""

    @ObservationIgnored
    private var loaded = false
    @ObservationIgnored
    private var disclosureTask: Task<Void, Never>?

    /**
     Web's `useEffect(…, [canSave])`, plus the disclosure subscription.

     The hint load is guarded so returning to Settings does not refetch: it is a
     masked string that only changes when this section itself changes it. The
     stream is guarded on its own task handle instead, the same way
     `RecurringViewModel` does — a live `db.watch()` should outlive one
     appearance of the view, not be torn down and stood back up on every one.
     */
    func start() async {
        await loadOnce()
        guard disclosureTask == nil else { return }
        disclosureTask = Task { [weak self] in
            guard let self else { return }
            do {
                for try await rows in try self.upiRepository.watchDisclosures() {
                    self.disclosures = rows
                }
            } catch {}
        }
    }

    private func loadOnce() async {
        guard !loaded else { return }
        loaded = true

        // Bound in two steps, exactly as `SettingsViewModel.loadSession()` does
        // it: `client.auth.session` is an async throwing property, and chaining
        // `.user` off a `try? await` of it is the kind of thing the compiler
        // blames on the wrong expression.
        guard let session = try? await client.auth.session else {
            canSave = false
            loading = false
            return
        }
        let user = session.user
        username = user.userMetadata["username"]?.stringValue ?? ""
        let allowed = !user.isAnonymous
        canSave = allowed
        guard allowed else {
            loading = false
            return
        }
        // The repository already degrades to the cached hint on a failed read,
        // so there is nothing to catch and nothing worth an error banner over a
        // field the user has not touched yet.
        hint = await upiRepository.getMyPaymentHandle()
        loading = false
    }

    /// Web's `save()`. `normalizedVpa` is already trimmed and lower-cased.
    ///
    /// Returns true when the handle was stored, so the view can clear its input
    /// — web does that with `setValue("")` inside the same `try`. The field
    /// stays in the view rather than moving up here: the section is otherwise
    /// stateless, and a bound `String` in an `@Observable` model would fight
    /// SwiftUI's own `@State` on the text field.
    @discardableResult
    func save(normalizedVpa: String) async -> Bool {
        if busy { return false }
        busy = true
        error = nil
        defer { busy = false }
        do {
            hint = try await upiRepository.savePaymentHandle(vpa: normalizedVpa, displayName: username)
            return true
        } catch let handleError as UpiHandleError {
            error = handleError.message
            return false
        } catch {
            // `self.` is load-bearing: the catch binding is also called `error`
            // and shadows the property.
            self.error = error.localizedDescription
            return false
        }
    }

    /// Web's `forget()`.
    func forget() async {
        if busy { return }
        busy = true
        error = nil
        defer { busy = false }
        do {
            try await upiRepository.forgetPaymentHandle()
            hint = nil
        } catch let handleError as UpiHandleError {
            error = handleError.message
        } catch {
            // `self.` is load-bearing here too -- see save(normalizedVpa:).
            self.error = error.localizedDescription
        }
    }

    /// The watch is a long-lived stream; without this it outlives the panel
    /// and keeps a strong `self` for the life of the process. Android's side
    /// is scoped by `WhileSubscribed(5000)` and needs no equivalent.
    deinit { disclosureTask?.cancel() }
}
